Update sentence selection model to use dynamic padding

matt-gardner opened the pull request:
Currently getting a weird error on model loading, but after that's fixed, this will hopefully just work as-is.

matt-gardner commented:
This is now ready to review.  It's updating the sentence selection model to be compatible with the new dynamic padding stuff that I've been working on.  This should make training these models a lot faster.  @haniesedghi, FYI.

pdasigi commented on deep_qa/layers/wrappers/add_encoder_mask.py:
```diff
@@ -0,0 +1,55 @@
+    (using word indices instead of embeddings for simplicity), which has been padded to be three
+    sentences, even though only two of them are actually used.  After passing it though an encoder,
+    you'll have something like [[vector], [vector], [vector]], and you want a mask that looks like
+    [1, 1, 0].  Keras' LSTMs and such won't give this to you.  This method adds it back.
```
You may want to say that you are assuming the encoder does not return sequences.

matt-gardner replied:
I updated the docstring in a couple of places to hopefully make this more clear.

pdasigi commented on deep_qa/layers/wrappers/add_encoder_mask.py:
```diff
@@ -0,0 +1,55 @@
+    """
+    def __init__(self, **kwargs):
+        # The base layer class overwrites self.supports_masking,
+        # so we add it afterwards.
```
Not sure what you meant by this comment. Did you mean to change the `supports_masking` field?

matt-gardner replied:
Thanks for catching that.  It's a copy-paste comment that was still in `OutputMask` because I forgot to remove it when I made it inherit from `MaskedLayer`.  Removed here and in `OutputMask`.

pdasigi commented on tests/layers/wrappers/add_encoder_mask_test.py:
```diff
@@ -0,0 +1,37 @@
+        ])
+        expected_mask = numpy.asarray([[0, 1, 0, 1, 1, 0, 1, 1]])
+        actual_mask = model.predict([test_background])
+        numpy.testing.assert_array_almost_equal(expected_mask, actual_mask)
```
Why not `assert_array_equal`?

matt-gardner replied:
Yeah, that's a good point.  These are boolean tensors, so equal should work.

matt-gardner commented:
Ok, I fixed those comments, @pdasigi, and I've run some timing benchmarks.  Without dynamic padding, I was truncating to 40 words per sentence.  The max sentence length is ~260 words, though I don't know the average sentence length.  One epoch with this truncation took about 45 minutes.  When I switched to dynamic padding I removed all truncation.  It was on track to finish in about 1.25 hours per epoch, but it ran out of memory on a batch with really long sentences.

If we assume a linear relationship between sentence length and runtime (which may not be accurate), we'd get maybe ~3 hours per epoch without any truncation (though it also would run out of memory, so we'd have to decrease the batch size, which would make it slower).  Dynamic padding, then, gives us a ~2-3x speedup in this model.
